File: src/test/java/kr/co/bookand/backend/policy/service/PolicyServiceTest.java
package kr.co.bookand.backend.policy.service;

import kr.co.bookand.backend.AccountTemplate;
import kr.co.bookand.backend.account.domain.Account;
import kr.co.bookand.backend.account.repository.AccountRepository;
import kr.co.bookand.backend.account.service.AccountService;
import kr.co.bookand.backend.policy.domain.Policy;
import kr.co.bookand.backend.policy.repository.PolicyRepository;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.Optional;

import static kr.co.bookand.backend.policy.domain.dto.PolicyDto.*;
import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.BDDMockito.given;

@ExtendWith(MockitoExtension.class)
class PolicyServiceTest {

    @Mock
    AccountService accountService;

    @Mock
    PolicyRepository policyRepository;

    @InjectMocks
    PolicyService policyService;


    @Test
    @DisplayName("정책 조회")
    void getPolicy() {
        // given
        PolicyRequest policyRequest = PolicyMockUtils.createPolicyRequest();
        Policy policy = policyRequest.toPolicy();
        given(policyRepository.findById(any())).willReturn(Optional.of(policy));

        //when
        PolicyResponse policyResponse = policyService.getPolicy(1L);

        //then
        assertEquals(policyResponse.title(), policyRequest.title());
        assertEquals(policyResponse.content(), policyRequest.content());
    }

    @Test
    @DisplayName("정책 생성")
    void createPolicy() {
        // given
        Account account = AccountTemplate.makeAccount1();
        given(accountService.getCurrentAccount()).willReturn(account);

        PolicyRequest policyRequest = PolicyMockUtils.createPolicyRequest();
        Policy mockPolicy = PolicyMockUtils.getMockPolicy();

        given(policyRepository.save(any())).willReturn(mockPolicy);

        //when
        PolicyResponse policyResponse = policyService.createPolicy(policyRequest);

        //then
        assertEquals(mockPolicy.getTitle(), policyResponse.title());
        assertEquals(mockPolicy.getContent(), policyResponse.content());
    }

    @Test
    @DisplayName("정책 수정")
    void updatePolicy() {
        // given
        Account account = AccountTemplate.makeAccount1();
        given(accountService.getCurrentAccount()).willReturn(account);

        PolicyRequest policyRequest = PolicyMockUtils.createPolicyRequest();
        Policy policy = policyRequest.toPolicy();
        given(policyRepository.findById(any())).willReturn(Optional.of(policy));

        //when
        PolicyResponse policyResponse = policyService.updatePolicy(policy.getId(), policyRequest);

        //then
        assertEquals(policyResponse.title(), policyRequest.title());
        assertEquals(policyResponse.content(), policyRequest.content());
    }

    @Test
    @DisplayName("정책 삭제")
    void removePolicy() {
        // given
        Account account = AccountTemplate.makeAccount1();
        given(accountService.getCurrentAccount()).willReturn(account);

        PolicyRequest policyRequest = PolicyMockUtils.createPolicyRequest();
        Policy policy = policyRequest.toPolicy();
        given(policyRepository.findById(any())).willReturn(Optional.of(policy));

        //when
        policyService.removePolicy(1L);

        //then
        assertTrue(true);
    }

}

File: src/test/kotlin/kr/co/bookand/backend/PolicyServiceTest.kt
package kr.co.bookand.backend

import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.matchers.shouldBe
import io.mockk.every
import io.mockk.mockk
import kr.co.bookand.backend.account.model.Account
import kr.co.bookand.backend.account.model.AccountStatus
import kr.co.bookand.backend.account.model.Role
import kr.co.bookand.backend.policy.model.Policy
import kr.co.bookand.backend.policy.dto.CreatePolicyRequest
import kr.co.bookand.backend.policy.repository.PolicyRepository
import kr.co.bookand.backend.policy.service.PolicyService
import java.time.LocalDateTime
import java.util.*

class PolicyServiceTest : BehaviorSpec({
    val policyRepository = mockk<PolicyRepository>()
    val policyService = PolicyService(policyRepository)

    Given("policy Test") {

        val account = Account(
            1L,
            "email",
            "password",
            "name",
            "provider",
            "providerEmail",
            "profileImage",
            LocalDateTime.now(),
            LocalDateTime.now(),
            null,
            Role.USER,
            AccountStatus.NORMAL
        )

        val adminAccount = Account(
            2L,
            "admin@email.com",
            "password",
            "admin",
            "provider",
            "providerEmail",
            "profileImage",
            LocalDateTime.now(),
            LocalDateTime.now(),
            null,
            Role.ADMIN,
            AccountStatus.NORMAL
        )

        val policyId = 1L
        val policy = Policy(
            1L,
            "Privacy Policy",
            "This is privacy policy",
            "This is privacy policy content"
        )

        val policy2 = Policy(
            2L,
            "Privacy Policy2",
            "This is privacy policy2",
            "This is privacy policy content2"
        )
        every { policyRepository.findById(1) } returns Optional.of(policy)
        When("policy 업데이트") {
            val request = CreatePolicyRequest(
                "Privacy Policy",
                "This is privacy policy",
                "This is new privacy policy content"
            )
            val updatedPolicy = policyService.updatePolicy(adminAccount, policyId, request)
            Then("it should update the content of the policy") {
                updatedPolicy.id shouldBe 1L
            }
        }
        When("policy 검색 by title") {
            every { policyRepository.findByName("This is privacy policy") } returns policy
            val result = policyService.getTitlePolicy("This is privacy policy")
            Then("it should return the policy with the given id") {
                result.policyId shouldBe policy.id
                result.name shouldBe policy.name
                result.content shouldBe policy.content
                result.title shouldBe policy.title
            }
        }
        When("policy 검색 by name") {
            every { policyRepository.findByName(policy.name) } returns policy
            val result = policyService.getTitlePolicy(policy.name)
            Then("it should return the policy with the given name") {
                result.policyId shouldBe policy.id
                result.name shouldBe policy.name
                result.content shouldBe policy.content
            }
        }
        When("policy 생성") {
            val request = CreatePolicyRequest(
                "This is privacy policy2",
                "Privacy Policy2",
                "This is privacy policy content2"
            )
            every { policyRepository.findByName("Privacy Policy2") } returns null
            every { policyRepository.save(any()) } returns policy2
            val result = policyService.createPolicy(adminAccount, request)

            Then("it should create the policy") {
                result.id shouldBe 2L
            }
        }
        When("policy 삭제") {
            every { policyRepository.findById(policyId) } returns Optional.of(policy)
            every { policyRepository.delete(policy) } returns Unit
            policyService.removePolicy(adminAccount, policyId)
            Then("it should remove the policy") {
                every { policyRepository.findById(policyId) } returns Optional.empty()
                policyRepository.findById(policyId) shouldBe Optional.empty()
            }
        }
    }
})